Readme.txt
=== Post Notification ===
Contributors: morty
Donate link: http://pn.xn--strbe-mva.de/forum.php?req=thread&id=4
Website: http://pn.xn--strbe-mva.de
Tags: notification, email, mail, post, subscribe, subscribe2, german, deutsch, french
Requires at least: 2.2
Tested up to: 2.7
Stable tag: 1.2.33


With each new post an email is sent to every registered User in the Database. The email can be text or HTML.

== Description ==
Post Notification 1.2 is **not compatible with WP < 2.2**. Please use Post Notification 1.1.x. You can find it, if you follow the "other versions" link
in the FYI-box on the right.  

* **[Changelog](http://dev.wp-plugins.org/browser/post-notification/trunk/post_notification/changelog.txt?format=txt)**
* **[Release Notes](http://pn.xn--strbe-mva.de/forum.php?req=main&id=1)**

Features: (**new in 1.2**) 

* **Support for WP 2.3**
* Can handle thousands of subscribers. **And has a simple back end to manage them.**
* **Subscribers can choose categories.**
* Easy to translate:The frontend should take a few minutes, the backend a bit longer. Please send me your translations! 
* The Post can be included, if you wish, only up to the more-tag.
* Mails can be sent as HTML or text.
* If you send text mails, the post is formated.
* Double Opt-in
* Frontend and mails are configured via templates. It is easy to change everything without messing in the code.
* Easy im- and export of Emails
* Every recipient gets his personal mail, with a link to change his subscribed categories or to unsubscribe.
* Nervous Finger option (Mail isn't sent right away - so you still have time to change things once you posted.)
* You can decide how many mails are sent in a burst and how long to pause between bursts.
* Decide on a per post basis whether to send a mail, or not, or just use the default.
* Captchas
* **Integrates into almost any theme without tweaks.**
* Translations: Deutsch (German), French (More welcome! Front end translations or modifications, too!)
* Does not need WP-Cron or anything like that.
* **Possebility to adjust mails with userfunctions**
* much more

= Changes in the templates since 1.1 =

* **select.tmpl** has been added.
* **activated.tmpl** has been removed.
* **strings.php** has five new strings.
* **unsubscribe.tmpl** now needs a @@conf_url.
* Mail templates can now use @@author. And @@unsub changed to @@conf_url.
* **subscribe.tmpl** need @@vars in the form.



= Where is the difference to Subscribe2? =

Subscribe2 as well as Post Notification are based on the same plugin. While Subscribe2 has it's emphasis on a
nice user interface and is more easy to configure this plugin is also suitable for professional sites who want
to send several thousand mails as day. There are two main reasons for this:

1. PN sends a "personal" email to each subscriber with a link to change his settings.
2. A subscriber can choose categories without having to register a Wordpress user.


== Installation ==
IMPORTANT: If you have an earlier version installed do not deactivate it. Just overwrite it.

1. Open the zip.
2. go into the post-notification directory.
3. Copy the post_notification (Underline!!) dir into you plugin dir.
4. You can have a look at the pictures if you want. Unfortunately they are packed into the file automatically

Activate the plugin, configure it carefully (options->Post Notification) and off you go.
For support please see the Forum: http://pn.xn--strbe-mva.de/
An please report problems. I can only fix stuff if I know that they are broken.

= If you are updating =
You might get errors, because old translations are not working. You can delete those dirs. (ru_RU, etc)

You have to update your old templates:

* **select.tmpl** has been added.
* **activated.tmpl** has been removed.
* **strings.php** has five new strings.
* **unsubscribe.tmpl** now needs a @@conf_url.
* Mail templates can now use @@author. And @@unsub changed to @@conf_url.
* **subscribe.tmpl** need @@vars in the form.

== Frequently Asked Questions ==
See the Forum : http://pn.xn--strbe-mva.de/


== Screenshots ==

1. Lots of configuration options
2. Decide whether to send a notification on a per post basis.
3. Filter the list of subscribers.
4. Manage the emails. You can add, remove and change the settings of a list of email addresses.



post_notification/changelog.txt
// 10.03.09
Lots of work done on the frontend. Commentsubscription seems to work now.
Backend: Some bugs in the management were fixed.

// 26.03.09
It is now possible to select the profile on the test-page.

// 25.03.09
admin_test seems to work. Let's get the translation done and release an alpha.

// 20.01.09
admin_list should be working. Some cleanup and the beta is ready! :-)

// 18.01.09
admin_list rewritten; still needs testing.


// 17.01.09
admin_manage rewritten; still needs testing -> database adjustments



// >>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> Changlog for 2.0 starts here <<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<<


2.0 has now reached an alpha state. It's therefore time to
start the changelog again.



Subscriber's IP is saved


//Release 1.2.11

// 02.01.2007
Bug: Testmail was sent although nosend was checked.
Bug: 50 mails are listed instead of 1.
It is now possible to mark all Emails listed.

// Release 1.2.10

// 02.01.2007
Better handling of the sidebar-snippet.

// 26.12.2007
Should be rfc 2047 & 822 compliant now

// Release 1.2.9

// 02.10.07
Updated French
Added some more abstraction between pn_send and admin_test by addin pn_sendmail

// 27.11.07
Bug: Removed empty line at the beginning of install.php.
 
// 27.11.07
Bug: When saving there was no output on the frontend.

// 23.11.7
Typo in settings

// 20.11.07
Added nl_NL by Joop Beris

// Release 1.2.8

// 15.11.07 
Improved Text formating
Test can be run without sending a mail.

// 14.11.07
Removed bug where @@p_n_title was not replaced in the <title> - Tag

// 14.11.07
Added an option to show empty categories

// 12.11.07 
Tried to fix the @@date - Problem. Once again. :-(

// Release 1.2.7

// 11.11.07 
Fixed small problem with post_notification_get_mailurl using index.php?

// 03.11.07
Reorganized the settings a bit.
Removed some typos in the settings
Added an option to remove unwanted filters.

// Release 1.2.6

// 25.10.07
More Debugging information

// 17.10.07
Removed dependency of mb_detect_encoding

// RELEASE 1.2.5

// 12.10.07
Bug: Empty unsubscribe mail was sent. (Thanks to Björn)
Bug: French translations didn't really work. Should be fixed now.
Moved Mailfix into it's own Plugin: http://wordpress.org/extend/plugins/wordpress-22-mailfix/
Bug: Typos in German translation.

// 11.10.07
Fixed problem with profile checking.

// RELEASE 1.2.4

// 07.10.07
Added change log to admin panel
Moved mailfix into the post_notification dir, to make the installation easier.
Small adjustment in the install.php: Only load install-helper.php it it needed - didn't solve the problem, but does no harm.
Also had some trouble with the svn-dir. This should be solved now.
Fixed some bugs in the documentation. 

// 05.10.07
Some variable cleanup in the frontend.php after the last bug fix. Nothing serious.

// RELEASE 1.2.3

// 05.10.07
Bug: Errors were not printed in the front end.

// RELEASE 1.2.2

// 02.10.07
Removed some warnings.
Removed possible Bug when subscribing.
Removed some small bugs which PHP fixed by itself.

// RELAESE 1.2.1

// 30.09.07
Bug removed a ; that found it's way behind a <?php in the admin_manage.php
Added French translation. Thanks to Michel Scriban 

// RELEASE 1.2.0

// 22.09.07
Added debugging information
Bug: It wasn't possible to turn debugging on
Works with WP 2.3

// 22.09.07
Solved some problems with incompatible mysql versions.
Added indexes to tables. (I lerned a lot about databases in the last few day - more optimizations to come)

// 16.09.07
Bug: Subscription didn't work because of a bug in a sql-query
Bug: POST_NOTIFICATION_PATH_URL was missing a "/"

// 13.09.07
Added database locking.

// 10.09.07
Remove all references with the path and replaced them with defines.

// 09.09.07
Added a check, whether there is enaugh time to send Mails.

// 06.09.07
It is now possible to send test-mails.

// 30.08.07
Added the Time-adjustment stuff from Karsten Tinnefeld

// 29.08.07
Lots of small changes. Thanks to Karsten Tinnefeld.
Added default categories. Thanks to Karsten Tinnefeld.
Fixed bug where the Plugin always got uninstalled when you deactivated it.
Installation is alsways run if you are in the admin_info. You don't get error messages when 
Chenged to automaticalle do the replacements

// 01.08.07
Bug: Klicking on the Link you were not aktivated until you confirmed your cat selection.

// 17.07.07
Fixed Problem with mysql not accepting '0' as default value for date_time.
User who did not Confirm are listed correctly now.

// 10.07.07
Fixed bug when selecting cats in the listing.

// 10.07.07
Bug: Couldn't delete Pages > 50
Bug: Javascript didn't work in Backend
The links in the list Email are now opened in a new window.

// 10.07.07
Replaced user_nicename by displayname for the Sting user for tha author.
Chenged the way Emails are listed. It's confirmed mails or unconfirmed.

// 08.07.07
Lots of small changes
Added select.tmpl
Added some dontion-Link for people > 500 subscribers
Updated doc
It's now easier to manage cats
Captcha class should be less easy to hack
Frontend cleanup
Wrote JS for category selection
added subscribe_snippet.php

// 01.07.07
Prepared inclusion of tinyMCE
Adjusted Backend for work with cats
Fixed bug sending mails: sending multiple mails
Redesigned frontend - Changing cats only works with link in Mail
Removed activated.tmpl
Adjusted templates

// 29.06.07
Fixed a bug subscribing. Should work again.

// 27.06.07
Rewrote the FE. Should be working now.
Fixed problem with umlauts. 
Sending mails should work correctly, too.

// 24.06.07
Added ignore_user_abort

// 24.06.07
Moved menu to the options menu.
Updated instructions.
Captcha didn't clean up.
Added Subscribed categories to the list (PN-Options)

// 23.06.07

post_notification/admin_info.php
<?php

#------------------------------------------------------
# INFO
#------------------------------------------------------
# This is part of the Post Notification Plugin for 
# Wordpress. Please see the Readme2.txt for details.
#------------------------------------------------------



function post_notification_admin_sub(){

	global $wpdb, $wp_filesystem;
	$t_posts = $wpdb->prefix . 'post_notification_posts';
	$t_subs = $wpdb->prefix . 'post_notification_subs';
	$t_emails = $wpdb->prefix . 'post_notification_emails';
	$datestr = get_option('date_format') . ' ' . get_option('time_format');
	
	//Run the install
	
	require_once(POST_NOTIFICATION_PATH . 'install.php');
	

	post_notification_install();

	
	if (is_object($wp_filesystem) ){
		
		echo "WP_FILESYSTEM!";
		$dir_handle=opendir(POST_NOTIFICATION_PATH);
		while (false !== ($file = readdir ($dir_handle))) {
			if(is_dir(POST_NOTIFICATION_PATH . $file) && $file[0] != '.' && $file[0] != '_') {
				if(is_wp_error(copy_dir(POST_NOTIFICATION_PATH . $file, POST_NOTIFICATION_DATA . $file))){
					echo '<div class="error">' . __('Failed to copy folder:', 'post_notification') .'<br/>';
					echo POST_NOTIFICATION_PATH . $file. ' -> ' . POST_NOTIFICATION_DATA . $file . '<br/>';
					echo __('Please ensure the Folder is writabel (777):', 'post_notification') .' ' .POST_NOTIFICATION_DATA. '</div>';
				} else {
					echo '<div class="error">' . __('Copied Folder:', 'post_notification') .'<br/>';
					echo POST_NOTIFICATION_PATH . $file. ' -> ' . POST_NOTIFICATION_DATA . $file . '</div>';
					if(delete(POST_NOTIFICATION_PATH . $file) === false){
						echo '<div class="error">' . __('Error deleting folder:', 'post_notification') . ' ' . POST_NOTIFICATION_PATH . $file . '</div>';
					}
				}
				
			}	
		}
		closedir($dir_handle); 
	}
	

	

	
	//Check for problems
	$mutex = @fopen(POST_NOTIFICATION_PATH . '_temp/post_notification.lock', 'w');
	if($mutex == false){
		echo '<div class="error">' .  __('Couldn\'t create File:', 'post_notification') . ' ' . POST_NOTIFICATION_PATH . '_temp/post_notification.lock<br />'
			. __('Please assure the the path is writeable.', 'post_notification') . ' '. __('This is an error!', 'post_notification') . '</div>'	;
	} else {
		fclose($mutex);
	}
	
	
	
	if(!function_exists('iconv')){
		echo '<div class="error">' .  __('PHP was compiled without iconv-libs. This might cause severe trouble using UTF-8 with special chars or umlauts.', 'post_notification'). ' ' 
			. __('Most hosters support this, as it is a sensible php-extension.', 'post_notification') . ' '
			. '<a href="http://php.net/manual/ref.iconv.php">' . __('More information.', 'post_notification') 
			. '</a></div>';
	}
	if(!function_exists('mb_detect_encoding')){
		echo '<div class="error">' . __('This version of PHP does not support the function mb_detect_encoding. You might experience some trouble with non-ASCII characters.', 'post_notification'). ' '
			. __('Most hosters support this, as it is a sensible php-extension.', 'post_notification') . ' '
			. '<a href="http://php.net/manual/ref.mbstring.php">' . __('More information.', 'post_notification') 
			. '</a></div>';
	}
	
	if(!function_exists('html_entity_decode')){
		echo '<div class="error">' .  __('This version of PHP does not support the function html_entity_decode. Sending out text mails will not be possible.', 'post_notification'). ' ' 
			. __('Please contact your hoster.', 'post_notification')
			. '</a></div>';		
	}

	if('' != ($dest = post_notification_installtheme())){
		echo '<div class="error">' . __('Couldn\'t create File:', 'post_notification') . $dest . ' ' 
		. __('Please assure the the path is writeable.', 'post_notification') . ' ' 
		. __('You can also copy the file from the plugin directory manually.', 'post_notification')
		. '</div>';
	}
	
	
	if(get_option('post_notification_debug') == 'yes'){
		echo '<div class="error">' . __('PN is in debugging mode. This should only be on, if something isn\'t working correctly.', 'post_notification') . '</div>';
	}
	if(get_option('post_notification_uninstall') == 'yes'){
		require_once(POST_NOTIFICATION_PATH . "install.php");
		echo '<div class="error">' . __('It seems like you want to uninstall this Plugin. Please deactivate it or change the setting the options. ', 'post_notification') . '</div>';
	} else if($wpdb->get_var("SHOW TABLES LIKE '$t_subs'") == NULL){
		echo '<div class="error">' . __('It seems like there is some trouble creating the tables in the DB. If you\'re lucky there should be some errormessages on this page.', 'post_notification') . '</div>';
	}
	
	if(ini_get('max_execution_time') != 0 && ini_get('max_execution_time') < 15){
		echo '<div class="error">' . __('The maximum executiontime is very low.', 'post_notification') . ' '
			. '<a href="http://php.net/manual/ref.info.php#ini.max-execution-time">' . __('More information.', 'post_notification') 
			. '</a></div>';;
	}
	
	if( (get_option('db_version') > 4772)  && (get_option('db_version') < 6124) && (substr( get_option('post_notification_template'), -5) == '.html')){
		echo '<div class="error">' . __('There is a bug in WP handling HTML-Mails. Please make sure a mail-plugin is installed.', 'post_notification') . 
		' <a href="http://wordpress.org/extend/plugins/wordpress-22-mailfix/">Wordpress 2.2 Mailfix</a> </div>';
		
	}
	
	
	//clean up - just in case.
	//Probably there's something with better performance, but as long there are not
	// 10 000 emails in the db this shouldn't matter. 
	/* Not well enaugh tested -> Next version
	$wpdb->query("	DELETE $t_cats, t_emails 
					FROM $t_cats LEFT JOIN t_emails
					WHERE email_addr IS NULL ");
	*/

	//Try to send.
	if(get_option('post_notification_debug') == 'yes'){
		post_notification_send_check(true);
	}
		
	//------------ Some info.....
	$nummails = $wpdb->get_var("SELECT COUNT(*) FROM $t_emails WHERE gets_mail = 1");
	echo '<p>'. __('Number of Subscribers:', 'post_notification') . "<b> $nummails </b></p>";
	//------------ Some advertising ----
	if($nummails > 500){
		echo '<p><b>' . __('Looks like you are realy using this plugin.', 'post_notification') . 
		'</b> <a href="http://pn.xn--strbe-mva.de/forum.php?req=thread&postid=8">' . __('What about a donation?.', 'post_notification') . '</a></p>';
	}
	
	
	echo '<p>' . __('The time is:', 'post_notification'); 
	echo ' <b>' . post_notification_date_i18n_tz($datestr, time()) . '</b></p>'; //Can use i18n_time as it uses date
	
	echo '<p>' .  __('Posts to be notified.', 'post_notification');
	$posts = $wpdb->get_results(
	"SELECT id, post_title, post_date_gmt, notification_sent, post_status, date_saved " 
		. post_notification_sql_posts_to_send(true) .
		" ORDER BY post_date_gmt DESC");
	echo '<table>
		<tr class="alternate"><th>'
			. __('Post-ID', 'post_notification') . '</th><th>'
			. __('Post title', 'post_notification') . '</th><th>'
			. __('Publish-date', 'post_notification') . '</th><th>'
			. __('Save-date', 'post_notification') . '</th></tr>';
	
	if (empty ($posts)) {
		echo '<tr><td colspan="5" class="alternate">' . __('None queued', 'post_notification') . '</td></tr>';
 
 	} else { 
		foreach($posts as $post){
			echo  '<tr class="alternate"><td>'
				. $post->id . '</td><td>' 
				. $post->post_title . '</td><td>'
				. post_notification_date_i18n_tz($datestr, post_notification_mysql2gmdate($post->post_date_gmt)) . '</td><td>' 
				. post_notification_date_i18n_tz($datestr, post_notification_mysql2gmdate($post->date_saved)) . '</td></tr>';
		
		}
	}
	echo '</table></p>';
	
	
	echo '<p>' . __('Mails to send:', 'post_notification');
	
	$emails = $wpdb->get_results("SELECT email_addr, obj_id, post_title
			FROM $t_queue q JOIN $t_emails e ON q.email_id = e.id
				JOIN " . $wpdb->posts . "p ON q.email_id = p.id
			WHERE q.type = 0 AND e.state = 0 ORDER BY q.obj_id");
	
	echo '<table>
		<tr class="alternate"><th>'
			. __('Email', 'post_notification') . '</th><th>'
			. __('Post-ID', 'post_notification') . '</th><th>'
			. __('Post title', 'post_notification') . '</th><th>';

	if (empty ($posts)) {
		echo '<tr><td colspan="3" class="alternate">' . __('None queued', 'post_notification') . '</td></tr>';
 
 	} else { 
		foreach($emails as $email){
			echo  '<tr class="alternate"><td>'
				. $email->email_addr . '</td><td>' 
				. $post->obj_id . '</td><td>'
				. $post->post_title . '</td></tr>';
		
		}
	}
	echo '</table></p>';
	
	echo '<p>' . __('Mails with errors:', 'post_notification');
	
	$emails = $wpdb->get_results("SELECT email_addr, obj_id, post_title
			FROM $t_queue q JOIN $t_emails e ON q.email_id = e.id
				JOIN " . $wpdb->posts . "p ON q.email_id = p.id
			WHERE q.type = 0 AND e.state = -1 ORDER BY q.obj_id");
	
	echo '<table>
		<tr class="alternate"><th>'
			. __('Email', 'post_notification') . '</th><th>'
			. __('Post-ID', 'post_notification') . '</th><th>'
			. __('Post title', 'post_notification') . '</th><th>';

	if (empty ($posts)) {
		echo '<tr><td colspan="3" class="alternate">' . __('None', 'post_notification') . '</td></tr>';
 
 	} else { 
		foreach($emails as $email){
			echo  '<tr class="alternate"><td>'
				. $email->email_addr . '</td><td>' 
				. $post->obj_id . '</td><td>'
				. $post->post_title . '</td></tr>';
		
		}
	}
	echo '</table></p>';
	
	
	echo '<p>' . __('The next mail will be sent:', 'post_notification') . ' ';
	if(get_option('post_notification_nextmove') == '-1'){
		 _e('None queued', 'post_notification');
	} else {
		echo post_notification_date_i18n_tz($datestr, get_option('post_notification_nextmove'));
	}
	echo '</p>';
	
	echo '<p>' . __('The last mail was sent:', 'post_notification') . ' ';
	echo post_notification_date_i18n_tz($datestr, get_option('post_notification_lastsend'));
	echo '</p>';
	echo '<p>' . __('The last post was saved:', 'post_notification') . ' ';
	echo post_notification_date_i18n_tz($datestr, get_option('post_notification_lastpost'));
	echo '</p>';

	
	
}
?>

post_notification/de_DE/fe_manage.php
<?php
/**
 * post_notification_profile_manage() will be called when a user is loged in.
 * $param is whatever link is passed to post_notification_fe_make_link($param).
 * p_n_p_manage() returns an array. It contains two elements: 'body' and 'header'
 * All Variables are replaced passing the Variables back to the super function.
 * The usage can be siplified by using post_notification_ldfile which loads the
 * first line of a file into 'header' and the rest into 'body'
 */


function post_notification_profile_manage($param){
	
	if($param == 'settings'){
		$content = post_notification_ldfile("man_settings.tmpl");
	} else if($param == 'post_cat'){
		$content = post_notification_ldfile("man_post_cat.tmpl");
	} else if($param == 'com_cat'){
		$content = post_notification_ldfile("man_com_cat.tmpl");
	} else if($param == 'com_post'){
		$content = post_notification_ldfile("man_com_post.tmpl");
	} else {
		$content = post_notification_ldfile("man_info.tmpl");
	}
	
	$body = &$content['body'];

	$navbar = '<a href="'. post_notification_fe_make_link('info') . '">Info</a> &middot; ' .
		'<a href="'. post_notification_fe_make_link('settings') . '">Einstellungen</a> &middot; '.
		'<a href="'. post_notification_fe_make_link('post_cat') . '">Abonierte Kategorien</a> &middot; ' .
		'<a href="'. post_notification_fe_make_link('com_cat') . '">Abonierte Kommentare (Kategorie)</a> &middot; ' .
		'<a href="'. post_notification_fe_make_link('com_post') . '">Abonierte Kommentare (Post)</a>';
		
	$body = str_replace('@@navbar',$navbar,$body);
	return $content;
	
	
	
}
?>

post_notification/post_notification.php
<?php

/*
Plugin Name: Post Notification
Plugin URI: http://pn.xn--strbe-mva.de/
Description: Sends an email to all subscribers. See readme or instructions for details.
Author: Moritz Str&uuml;be
Version: 2.0.b.1
License: GPL
Author URI: http://xn--strbe-mva.de
Min WP Version: 2.3

*/

#------------------------------------------------------
# INFO
#------------------------------------------------------
# This is part of the Post Notification Plugin for 
# Wordpress. Please see the Readme2.txt for details.
#------------------------------------------------------

/**
 * This file has all the stuff that is really needed to initialize the plugin.
 */



define("POST_NOTIFICATION_PLUGIN_DIR", dirname(plugin_basename(__FILE__)));
define("POST_NOTIFICATION_PATH_REL", PLUGINDIR . '/' . POST_NOTIFICATION_PLUGIN_DIR);
define("POST_NOTIFICATION_PATH",  ABSPATH . POST_NOTIFICATION_PATH_REL . '/');
define("post_notification_path",  ABSPATH . POST_NOTIFICATION_PATH_REL . '/'); //To fix some problems.
define("POST_NOTIFICATION_PATH_URL",   get_option('siteurl') . '/wp-content/plugins/'. POST_NOTIFICATION_PLUGIN_DIR. '/');
define("POST_NOTIFICATION_DATA", WP_CONTENT_DIR . '/post-notification/');

//Include all the helper functions
require_once(POST_NOTIFICATION_PATH . "functions.php");

/**
 * This function returns the header of Post Notification as a string
 */
function post_notification_feheader(){
	require_once(POST_NOTIFICATION_PATH . 'frontend.php');
	$content = post_notification_page_content();
	return $content['header'];
}

/**
 * This function returns the body of Post Notification as a string
 */
function post_notification_febody(){
	require_once(POST_NOTIFICATION_PATH . 'frontend.php');
	$content = post_notification_page_content();
	return $content['body'];
}


/// Add the Admin panel
function post_notification_admin_adder(){ 
	$name = add_options_page('Post Notification','Post Notification', 8, 'post_notification/admin.php', 'post_notification_admin');
	
	//This is for future use.
	//add_action('load-' . $name, 'post_notification_admin_load');
}


/// Show the admin panel
function post_notification_admin(){
	require_once (POST_NOTIFICATION_PATH . "admin.php");
	post_notification_admin_page();
}

/// For Future use
function post_notification_admin_load(){
	require_once (POST_NOTIFICATION_PATH . "admin.php");
	post_notification_admin_page_load();
}


/// Add the subscribe-page to the meta-information
function post_notification_meta(){
	if(get_option('post_notification_page_meta') == 'yes'){
		$link = post_notification_get_link();
		if($link){
			echo '<li><a href="' . $link. '">' . get_option('post_notification_page_name') . '</a></li>';
		}
	}
}





/// Add the option to whether to send a notification
function post_notification_form() {
	global $post_ID, $post, $wpdb;
	$t_posts = $wpdb->prefix . 'post_notification_posts';
	
	load_plugin_textdomain('post_notification', POST_NOTIFICATION_PATH_REL);
	
	$textyes = __('Yes', 'post_notification');
	$textdef = __('Default', 'post_notification');
	$default = false;
	
	if(0 != $post_ID){ //We've got an ID.
		$status= $wpdb->get_var("SELECT notification_sent FROM $t_posts WHERE post_ID = '$post_ID'");

		if(isset($status)){ //It's in the DB
			if($status = 0){ //It will be sent in the future
				$default = true;
				$textdef = __('Send Mails in queue.', 'post_notification');
			} else { //It has been sent or is not being sent.
				$sendN = 'selected="selected"';
				if($status = 1){ //If it's 1 it has already been sent before
					$textyes = __('Resend', 'post_notification');
				}
			}
		} else { //This one has been written bevore PN was installed.
				$sendN = 'selected="selected"';
		}
	} else {
		$default = true;
	}
	
	if (!function_exists('add_meta_box')){
	?>
	<div id="advancedstuff" class="dbx-group">
		<div class="dbx-b-ox-wrapper">
			<fieldset id="emailnotification" class="dbx-box">
			<div class="dbx-h-andle-wrapper"><h3 class="dbx-handle">Post Notification</h3></div>
				<div class="dbx-c-ontent-wrapper">
					<div class="dbx-content">
	<?php } 
						_e('Send notification when publishing?', 'post_notification'); ?>
						<select id="post_notification_notify" name="post_notification_notify">
							<?php if($default){?>
								<option value="def" selected="selected"><?php echo $textdef;  ?></option>
							<?php } ?>
							<option value="yes" <?php echo $sendY; ?>><?php echo $textyes;  ?></option>
							<option value="no"  <?php echo $sendN; ?>><?php _e('No', 'post_notification') ?></option>
						</select><?php 
						
						if (!function_exists('add_meta_box')){ ?>
						
					</div>
				</div>
			</fieldset>
	</div>
	<?php }
}

/**
 * Thus function is a wrapper for the function, which adds the checkbox
 * to subscibe to commens
 * @param $post_id
 * @return unknown_type
 */
function post_notification_comment_subscription_wrap($post_id){
	require_once(POST_NOTIFICATION_PATH . "frontend.php");
	post_notification_comment_subscription($post_id);

}

/**
 * This function handles when a comment is being sent.
 * @param $comment_ID
 * @return unknown_type
 */
function post_notification_com_handle($comment_ID){
	//Get the commnet
	$com = get_comment($comment_ID);
	//First check, wether the checkbox has been shown
	if(isset($_POST['pn_subscribe_hint'])){
		
		/// @todo Füllen
	}
	
	if($com->comment_approved != 1) return;
	require_once(POST_NOTIFICATION_PATH  . 'sendmail.php');
	post_notification_add_to_queue($comment_ID, 1);
	
}


/// Add a Post to the notificationlist
function post_notification_add($post_ID)  {
	global $wpdb;  
	$post = get_post($post_ID);
	
	$t_posts = $wpdb->prefix . 'post_notification_posts'; 
	$notify    = $_POST['post_notification_notify'];
	//Todo, userlevels 
	// if($notify != '')
	if($notify == '') $notify = 'def';
	$status = $wpdb->get_var("SELECT notification_sent FROM $t_posts WHERE post_ID = '$post_ID'");
	
	if($notify == 'def' && !isset($status)){ //default is not to change
		if(get_option('db_version')< 4772){
			if($post->post_status == 'post')		$notify = get_option('post_notification_send_default');
			if($post->post_status == 'private')		$notify = get_option('post_notification_send_default');
			if($post->post_status == 'static')		$notify = get_option('post_notification_send_page');
		} else {
			if($post->post_type == 'post')			$notify = get_option('post_notification_send_default');
			if($post->post_type == 'post' 
			  && $post->post_status == 'private')	$notify = get_option('post_notification_send_default');
			if($post->post_type == 'page')			$notify = get_option('post_notification_send_page');
		} 
	}
	
	
	
	if($notify == 'yes'){
		if(isset($status)) $wpdb->query("UPDATE $t_posts  SET notification_sent = 0 WHERE post_id = " . $post_ID  );
		else $wpdb->query("INSERT INTO $t_posts  (post_ID, notification_sent) VALUES ('$post_ID',  0)");
	} else if($notify == 'no'){
		if($status != -1){ //Mails are sent - no reason to change this
			if(isset($status)) $wpdb->query("UPDATE $t_posts  SET notification_sent = -1 WHERE post_id = " . $post_ID  );
			else $wpdb->query("INSERT INTO $t_posts  (post_ID, notification_sent) VALUES ('$post_ID',  -1)");
		}
	}
	// We should have an entry now, so lets write the time.
	$wpdb->query("UPDATE $t_posts  SET date_saved = '" . post_notification_date2mysql() . "' WHERE post_id = " . $post_ID  );
	post_notification_set_next_move(); 
}


/// Check whether a Mail is to be sent.
function post_notification_send_check($force = false){
	if(	(get_option('post_notification_nextmove') == -1 || //No articles to send 
			get_option('post_notification_nextmove') > time()) && //Or artivles in the future
		add_option('post_notification_emptyqueue') == 1) return; //AND the queue is empty.
	if((get_option('post_notification_debug') != 'yes') || $force){ //Don't send in debugmode.
		require_once(POST_NOTIFICATION_PATH  . 'sendmail.php');
		post_notification_send();
	}
}


/// A wrapper function for the installation
function post_notification_install_wrap(){
	require_once(POST_NOTIFICATION_PATH . 'install.php');
	post_notification_install();
}

/// A wrapper function for the deinstallation
function post_notification_uninstall_wrap(){
	require_once(POST_NOTIFICATION_PATH . 'install.php');
	post_notification_uninstall();
}

function post_notification_fe_check(){
	//post_notification_vardump($GET);
	if($_GET['POST_NOTIFICATION_FE_CHECK'] == '1'){
		//echo "\n\n---------------------------------------\n\n";
		post_notification_page_content();
	}
}


//********************************************//
// Actions
//********************************************//

function post_notification_gui_init(){

	if (function_exists('add_meta_box')) {
	//This starts with WP 2.5
		add_meta_box('post_notification', 'Post Notification', post_notification_form,'post', 'normal' );
		add_meta_box('post_notification', 'Post Notification', post_notification_form,'page', 'normal' );
		
	} else {
		// Notify box in advanced mode
		add_action('edit_form_advanced', 'post_notification_form', 5);
		// Notify box in page mode
		add_action('edit_page_form', 'post_notification_form', 5);
	}
	
	// Notify box in simple mode
	add_action('simple_edit_form', 'post_notification_form', 5);
}


add_action('init', 'post_notification_fe_check');

add_action('admin_menu', 'post_notification_gui_init');


// Admin menu
add_action('admin_menu', 'post_notification_admin_adder');


// Save for notification
add_action('save_post', 'post_notification_add', 5);

// Send the notification
add_action('shutdown', 'post_notification_send_check');
//add_action('admin_footer', 'post_notification_send');

// Trigger installation
add_action('activate_post_notification/post_notification.php','post_notification_install_wrap');

// Trigger uninstallation
add_action('deactivate_post_notification/post_notification.php','post_notification_uninstall_wrap');

// Add Metainformation
add_action('wp_meta', 'post_notification_meta', 0); 

// Copy template to theme
add_action('switch_theme', 'post_notification_installtheme');

// Add subscibe-link to Comment-Area
add_action('comment_form', 'post_notification_comment_subscription_wrap');


// Add comments to queue
add_action('comment_post', 'post_notification_com_handle');
add_action('wp_set_comment_status', 'post_notification_com_handle');



// Replacement of Post-Strings.
if(get_option('post_notification_filter_include') == 'yes'){
	require_once(POST_NOTIFICATION_PATH . 'frontend.php');
	add_filter('the_content', 'post_notification_filter_content');
	add_filter('the_title', 'post_notification_filter_content');
	add_filter('single_post_title', 'post_notification_filter_content');
}

?>
